test/PushNotificationTests/Shared.cpp

```diff
@@ -0,0 +1,155 @@
+            CreationCollisionOption::OpenIfExists).get();
+    }
+
+    StorageFile OpenDocFile(std::wstring filename)
+    {
+        return KnownFolders::DocumentsLibrary().GetFileAsync(filename).get();
+    }
+
+    wil::unique_handle Execute(const std::wstring& command, const std::wstring& args,
+        const std::wstring& directory)
+    {
+        SHELLEXECUTEINFO ei{};
+        ei.cbSize = sizeof(SHELLEXECUTEINFO);
+        ei.fMask = SEE_MASK_NOCLOSEPROCESS | SEE_MASK_DOENVSUBST;
+        ei.lpFile = command.c_str();
+        ei.lpParameters = args.c_str();
+        ei.lpDirectory = directory.c_str();
+
+        if (!ShellExecuteEx(&ei))
+        {
+            auto lastError = GetLastError();
+            VERIFY_WIN32_SUCCEEDED(lastError);
+        }
+
+        wil::unique_handle process{ ei.hProcess };
+        return process;
+    }
+
+    void RunCertUtil(const std::wstring& path, bool removeCert)
+    {
+        std::wstring action = (removeCert ? L"-delstore" : L"-addstore");
+        std::wstring args{ action + L" TrustedPeople " + path };
+        // VERIFY_ARE_NOT_EQUAL(args, args);
+        auto process = Execute(L"%SystemRoot%\\system32\\certutil.exe",
+            args.c_str(), g_deploymentDir);
+
+        // Wait for the cer to be installed.
+        auto waitResult = WaitForSingleObject(process.get(), c_phaseTimeout);
+        if (waitResult != WAIT_OBJECT_0)
+        {
+            auto lastError = GetLastError();
+            VERIFY_WIN32_FAILED(lastError);
+        }
+
+        // Make sure the exitcode for the tool is success.
+        DWORD exitCode{};
+        THROW_IF_WIN32_BOOL_FALSE(GetExitCodeProcess(process.get(), &exitCode));
+        VERIFY_ARE_EQUAL(exitCode, 0);
+    }
+
+    void InstallPackage(const std::wstring& packagePath)
+    {
+        // Deploy packaged app to register handler through the manifest.
+        PackageManager manager;
+        IVector<Uri> depPackagePaths;
+        auto result = manager.AddPackageAsync(Uri(packagePath), depPackagePaths,
+            DeploymentOptions::ForceApplicationShutdown).get();
+        auto errorText = result.ErrorText();
+        auto errorCode = result.ExtendedErrorCode();
+        VERIFY_SUCCEEDED(errorCode, errorText.c_str());
+    }
+
+    void UninstallPackage(const std::wstring& packageFullName)
+    {
+        // Deploy packaged app to register handler through the manifest.
+        PackageManager manager;
+        IVector<Uri> depPackagePaths;
+        auto result = manager.RemovePackageAsync(packageFullName).get();
+        auto errorText = result.ErrorText();
+        auto errorCode = result.ExtendedErrorCode();
+        VERIFY_SUCCEEDED(errorCode, errorText.c_str());
+    }
+
+    wil::unique_event CreateTestEvent(const std::wstring& eventName)
+    {
+        bool alreadyExists = false;
+        SECURITY_ATTRIBUTES attributes = {};
+        wil::unique_hlocal_security_descriptor descriptor;
+
+        // Grant access to world and appcontainer.
```

Comment:
World is too broad. Maybe okay for a test but bad in terms of setting a good example for others.

Reply:
This is only being used for testing and communication between TAEF / App.

Reply:
And then if this test is run on a server as part of CI, you have a potential vulnerability: Another test running on the same server under a different user can still access the event and mutate it, confusing your test.

Reply:
@pmpurifoy What's the reason for the accesses? AppContainer so the test app activated by the Push API can access? World because...? 

@oldnewthing do you have a suggested alternative design (something more invasive than tweaking the securitydescription)?

Here we have TestProcess creating another process via Push's API which wants to positively ACK to the TestProcess it was started appropriately as expected. Either TestProcess creates a resource which PushProcess taps (which we have here - CreateEvent then SetEvent, same diff if TestProcess creates a NamedPipe and PushProcess connects to it, etc) --or-- the PushProcess creates some resource which the TestProcess looks for (file, regkey, etc) but any such resource has the same security issue plus the durability problem.

Only option I can think of is if PushProcess creates something in ApplicationData e.g. TestProcess calls ApplicationDataManager.CreateForPackageFamily(blah).LocalFolder.ClearAsync() to clear any previous leftover data, then call Push API and sit in a polling loop looking for the ACK file. PushProcess when happy calls AppData.LocalFolder.CreateFileAsync() to create the ACK file. That avoids the security concern but introduces multiple undesired 'features'

I can think of alternative designs but nothing I'd really call _better_. But thought I'd ask, as this "Test does something to create a process which needs to ACK it's done and happy and success" pattern is quite common (and not just in ProjectReunion). I've never enjoyed troubleshooting those failures as test logs just show "I waited for X to happen and eventually gave up, but got me what went wrong". Would love to see a better answer (if possible).

Reply:
We can tighten the ACLs on the shared event so it is at least restricted to the same user. The rules for low IL access are

1. Check whether the app *would have* access if it were running at medium IL.
2. Check that the app has access via a capability, an explicit appid SID, or a wildcard "anybody at low IL" SID.

The code was using "World" as a lazy but extremely insecure way to get 1. It should do something that meets at least minimum security guidelines, say, by adding the current user SID. That way, the app can be attacked only by the same user, which is not an elevation.